File: Codes in C++/Array/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size();i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int l=i+1;int r=nums.size()-1;
            while(l<r){
                int sum=nums[i]+nums[l]+nums[r];
                if(sum>0)
                    r--;
                else if (sum<0)
                    l++;
                else if(sum==0){
                    vector<int> temp={nums[i],nums[l],nums[r]};
                    res.push_back(temp);
                    l++;
                    r--;
                
                while(l<r&&nums[l]==nums[l-1])l++;
                while(l<r&&nums[r]==nums[r+1])r--;
                }
            }
        }
        return res;
        }
};

File: Codes in Python3/Two Pointers/3Sum.py
#Using Two Pointer
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        nums.sort()
        res=[]
        for k in range(n):
            if nums[k]>0:
                break
            if k>0 and nums[k]==nums[k-1]:
                continue
            l=k+1
            r=n-1
            while l<r:
                totalSum=nums[l]+nums[r]+nums[k]
                if totalSum<0:
                    l+=1
                elif totalSum>0:
                    r-=1
                else:
                    res.append([nums[l],nums[r],nums[k]])
                    l+=1
                    r-=1
                    while l<r and nums[l]==nums[l-1]:
                        l+=1
                    while l<r and nums[r]==nums[r+1]:
                        r-=1

        return res
# TC=O(n^2)
# SC=O(1)
        